src/vulkan-renderer/wrapper/cpu_texture.cpp

```diff
@@ -0,0 +1,87 @@
+
+    spdlog::debug("Loading texture file {}.", file_name);
+
+    // Load the texture file using stb_image library.
+    // Force stb_image to load an alpha channel as well.
+    stbi_uc *texture_data = stbi_load(file_name.c_str(), &m_texture_width, &m_texture_height, nullptr, STBI_rgb_alpha);
+
+    if (texture_data == nullptr) {
+        spdlog::error("Could not load texture file {} using stbi_load!  Falling back to error texture.", file_name);
+        generate_error_texture_data();
+    } else {
+
+        // TODO: We are currently hard coding the number of channels with STBI_rgb_alpha.
+        //       Eventually, we probably need to pass this information into this class from
+        //       a higher level class - like a material loader class.
+        //       So, as an example, if the material loader is loading a normal map, we know
+        //       we need to tell this class to load a 3 channel texture.  If it can, great.
+        //       If it can not, then this class probably needs to load a 3 channel error texture.
+        m_texture_channels = 4;
+
+        // TODO: We are currently only supporting 1 mip level.
+        m_mip_levels = 1;
+
+        const int texture_data_size{m_texture_width * m_texture_height * m_texture_channels};
+        m_texture_data.resize(texture_data_size);
+        std::copy(texture_data, texture_data + texture_data_size, m_texture_data.begin());
+
+        spdlog::debug("Texture dimensions: width: {}, height: {}, channels: {} mip levels: {}.", m_texture_width,
+                      m_texture_height, m_texture_channels, m_mip_levels);
+    }
+
+    // Discard the texture data.
+    stbi_image_free(texture_data);
+}
+
+CpuTexture::CpuTexture(CpuTexture &&other) noexcept
+    : m_file_name(std::move(other.m_file_name)), m_name(std::move(other.m_name)),
+      m_texture_width(other.m_texture_width), m_texture_height(other.m_texture_height),
+      m_texture_channels(other.m_texture_channels), m_mip_levels(other.m_mip_levels),
+      m_texture_data(std::move(other.m_texture_data)) {}
+
+CpuTexture::~CpuTexture() {}
+
+void CpuTexture::generate_error_texture_data() {
+    m_texture_width = 512;
+    m_texture_height = 512;
+    m_texture_channels = 4;
+    m_mip_levels = 1;
+
+    // Create an 8x8 checkerboard pattern of squares.
+    const std::size_t square_dimension{64};
+    // pink, purple
+    constexpr std::array<std::array<unsigned char, 4>, 2> COLORS{{{0xFF, 0x69, 0xB4, 0xFF}, {0x94, 0x00, 0xD3, 0xFF}}};
+
+    const auto get_color = [](std::size_t x, std::size_t y, std::size_t square_dimension,
+                              std::size_t colors) -> std::size_t {
+        return (std::size_t(x / square_dimension) + std::size_t(y / square_dimension)) % colors;
+    };
+
+    m_texture_data.reserve(m_texture_width * m_texture_height * m_texture_channels);
+    for (std::size_t y = 0; y < m_texture_height; y++) {
+        for (std::size_t x = 0; x < m_texture_width; x++) {
```

Comment:
Is there a reason why you do `y` and then `x`? (just curious)

Reply:
Iterating through the height first can be more cache friendly.

Reply:
Yeah.

In a normal 1D to 2D mapping, you have normally something like this in 1D:

```
abcdefghi
```

that in 2D is represented as:

```
abc
def
ghi
```

So, if you do something like:

```
for x in width:
    for y in height:
        1d[x + y*width] = ...
```

that iterates in this order by index:

```
036147258
adgbehcfi
```

It sort of 'hops' around instead of going linearly across the memory.

But, for this particular texture, you are correct.  It does not really matter because the pattern of the texture is symmetrical across X and Y.

But, with an inner Y loop, the get_color function is building data down Y but the vector.insert is storing across X.

So, I don't know.  I think it is just a good habit for 1D to 2D array work - unless algorithmicaly there is a strong reason not to do it.